File: sdk/python/brocode_sdk.py
# BroCode Sentinel Python SDK
# Official Python client for BroCode Sentinel API

import requests
import json
import time
from typing import Dict, List, Optional, Any

class BroCodeSDK:
    """
    Official Python SDK for BroCode Sentinel API.
    
    Usage:
        from brocode_sdk import BroCodeSDK
        
        client = BroCodeSDK(
            api_key="bk_your_api_key",
            base_url="https://your-domain.com"
        )
        
        # Register user
        result = client.register_user("john", "english", keystrokes_data)
        
        # Verify user
        result = client.verify_user("john", "english", keystrokes_data)
        
        # Get trust score
        score = client.get_trust_score("john")
    """
    
    def __init__(self, api_key: str, base_url: str = "http://localhost:5000", timeout: int = 30):
        """
        Initialize BroCode SDK client
        
        Args:
            api_key: Your API key (starts with 'bk_')
            base_url: Base URL of BroCode Sentinel API
            timeout: Request timeout in seconds
        """
        self.api_key = api_key
        self.base_url = base_url.rstrip('/')
        self.timeout = timeout
        self.session = requests.Session()
        self.session.headers.update({
            'X-API-Key': api_key,
            'Content-Type': 'application/json'
        })
    
    def _request(self, method: str, endpoint: str, data: Optional[Dict] = None, params: Optional[Dict] = None) -> Dict:
        """Make HTTP request to API"""
        url = f"{self.base_url}/api/v1{endpoint}"
        
        try:
            response = self.session.request(
                method=method,
                url=url,
                json=data,
                params=params,
                timeout=self.timeout
            )
            
            response.raise_for_status()
            return response.json()
            
        except requests.exceptions.HTTPError as e:
            error_data = {}
            try:
                error_data = e.response.json()
            except:
                error_data = {"error": str(e)}
            
            raise BroCodeAPIError(
                message=error_data.get("error", "API request failed"),
                status_code=e.response.status_code,
                details=error_data
            )
        except requests.exceptions.RequestException as e:
            raise BroCodeAPIError(
                message=f"Request failed: {str(e)}",
                status_code=0,
                details={}
            )
    
    def health_check(self) -> Dict:
        """
        Check API health status
        
        Returns:
            Dict with health status
        """
        return self._request('GET', '/health')
    
    def register_user(self, username: str, language: str, keystrokes: List[Dict], email: Optional[str] = None) -> Dict:
        """
        Register a new user
        
        Args:
            username: Username
            language: Language (english, spanish, etc.)
            keystrokes: List of keystroke timing data
            email: Optional email address
        
        Returns:
            Registration result
        """
        data = {
            "username": username,
            "language": language,
            "keystrokes": keystrokes
        }
        
        if email:
            data["email"] = email
        
        return self._request('POST', '/auth/register', data=data)
    
    def verify_user(self, username: str, language: str, keystrokes: List[Dict]) -> Dict:
        """
        Verify user authentication
        
        Args:
            username: Username
            language: Language
            keystrokes: Keystroke timing data
        
        Returns:
            Verification result with trust score
        """
        data = {
            "username": username,
            "language": language,
            "keystrokes": keystrokes
        }
        
        return self._request('POST', '/auth/verify', data=data)
    
    def get_trust_score(self, username: str) -> Dict:
        """
        Get current trust score for user
        
        Args:
            username: Username
        
        Returns:
            Trust score and components
        """
        return self._request('GET', f'/users/{username}/trust')
    
    def get_user_sessions(self, username: str, limit: int = 10, offset: int = 0) -> Dict:
        """
        Get user's session history
        
        Args:
            username: Username
            limit: Number of sessions to return
            offset: Pagination offset
        
        Returns:
            List of sessions
        """
        params = {"limit": limit, "offset": offset}
        return self._request('GET', f'/users/{username}/sessions', params=params)
    
    def register_webhook(self, url: str, events: List[str], secret: Optional[str] = None) -> Dict:
        """
        Register a webhook for events
        
        Args:
            url: Webhook URL
            events: List of events to subscribe to
            secret: Optional webhook secret
        
        Returns:
            Webhook registration result
        """
        data = {
            "url": url,
            "events": events
        }
        
        if secret:
            data["secret"] = secret
        
        return self._request('POST', '/webhooks', data=data)
    
    def get_stats(self) -> Dict:
        """
        Get system statistics
        
        Returns:
            System statistics
        """
        return self._request('GET', '/stats')


class BroCodeAPIError(Exception):
    """Exception raised for API errors"""
    
    def __init__(self, message: str, status_code: int, details: Dict):
        self.message = message
        self.status_code = status_code
        self.details = details
        super().__init__(self.message)
    
    def __str__(self):
        return f"BroCodeAPIError({self.status_code}): {self.message}"


# Convenience functions

def create_client(api_key: str, base_url: str = "http://localhost:5000") -> BroCodeSDK:
    """
    Create a BroCode SDK client
    
    Args:
        api_key: Your API key
        base_url: Base URL of API
    
    Returns:
        BroCodeSDK client instance
    """
    return BroCodeSDK(api_key=api_key, base_url=base_url)


# Example usage
if __name__ == "__main__":
    # Initialize client
    client = BroCodeSDK(
        api_key="bk_your_api_key_here",
        base_url="http://localhost:5000"
    )
    
    # Health check
    try:
        health = client.health_check()
        print(f"API Status: {health['status']}")
    except BroCodeAPIError as e:
        print(f"Error: {e}")
    
    # Register user
    try:
        keystrokes = [
            {"key": "a", "press_time": 100, "release_time": 150},
            {"key": "b", "press_time": 200, "release_time": 250}
        ]
        
        result = client.register_user(
            username="john_doe",
            language="english",
            keystrokes=keystrokes,
            email="john@example.com"
        )
        
        print(f"Registration: {result}")
    except BroCodeAPIError as e:
        print(f"Error: {e}")
    
    # Get trust score
    try:
        score = client.get_trust_score("john_doe")
        print(f"Trust Score: {score['trust_score']}")
    except BroCodeAPIError as e:
        print(f"Error: {e}")

File: sdk/javascript/brocode_sdk.js
// BroCode Sentinel JavaScript/TypeScript SDK
// Official JavaScript client for BroCode Sentinel API

class BroCodeSDK {
    /**
     * Initialize BroCode SDK client
     * @param {string} apiKey - Your API key (starts with 'bk_')
     * @param {string} baseUrl - Base URL of BroCode Sentinel API
     */
    constructor(apiKey, baseUrl = 'http://localhost:5000') {
        this.apiKey = apiKey;
        this.baseUrl = baseUrl.replace(/\/$/, '');
        this.timeout = 30000; // 30 seconds
    }

    /**
     * Make HTTP request to API
     * @private
     */
    async _request(method, endpoint, data = null, params = null) {
        const url = new URL(`${this.baseUrl}/api/v1${endpoint}`);

        if (params) {
            Object.keys(params).forEach(key =>
                url.searchParams.append(key, params[key])
            );
        }

        const options = {
            method: method,
            headers: {
                'X-API-Key': this.apiKey,
                'Content-Type': 'application/json'
            }
        };

        if (data) {
            options.body = JSON.stringify(data);
        }

        try {
            const controller = new AbortController();
            const timeoutId = setTimeout(() => controller.abort(), this.timeout);

            options.signal = controller.signal;

            const response = await fetch(url.toString(), options);
            clearTimeout(timeoutId);

            const responseData = await response.json();

            if (!response.ok) {
                throw new BroCodeAPIError(
                    responseData.error || 'API request failed',
                    response.status,
                    responseData
                );
            }

            return responseData;
        } catch (error) {
            if (error instanceof BroCodeAPIError) {
                throw error;
            }
            throw new BroCodeAPIError(
                `Request failed: ${error.message}`,
                0,
                {}
            );
        }
    }

    /**
     * Check API health status
     * @returns {Promise<Object>} Health status
     */
    async healthCheck() {
        return await this._request('GET', '/health');
    }

    /**
     * Register a new user
     * @param {string} username - Username
     * @param {string} language - Language (english, spanish, etc.)
     * @param {Array} keystrokes - Keystroke timing data
     * @param {string} email - Optional email address
     * @returns {Promise<Object>} Registration result
     */
    async registerUser(username, language, keystrokes, email = null) {
        const data = {
            username,
            language,
            keystrokes
        };

        if (email) {
            data.email = email;
        }

        return await this._request('POST', '/auth/register', data);
    }

    /**
     * Verify user authentication
     * @param {string} username - Username
     * @param {string} language - Language
     * @param {Array} keystrokes - Keystroke timing data
     * @returns {Promise<Object>} Verification result with trust score
     */
    async verifyUser(username, language, keystrokes) {
        const data = {
            username,
            language,
            keystrokes
        };

        return await this._request('POST', '/auth/verify', data);
    }

    /**
     * Get current trust score for user
     * @param {string} username - Username
     * @returns {Promise<Object>} Trust score and components
     */
    async getTrustScore(username) {
        return await this._request('GET', `/users/${username}/trust`);
    }

    /**
     * Get user's session history
     * @param {string} username - Username
     * @param {number} limit - Number of sessions to return
     * @param {number} offset - Pagination offset
     * @returns {Promise<Object>} List of sessions
     */
    async getUserSessions(username, limit = 10, offset = 0) {
        const params = { limit, offset };
        return await this._request('GET', `/users/${username}/sessions`, null, params);
    }

    /**
     * Register a webhook for events
     * @param {string} url - Webhook URL
     * @param {Array<string>} events - List of events to subscribe to
     * @param {string} secret - Optional webhook secret
     * @returns {Promise<Object>} Webhook registration result
     */
    async registerWebhook(url, events, secret = null) {
        const data = {
            url,
            events
        };

        if (secret) {
            data.secret = secret;
        }

        return await this._request('POST', '/webhooks', data);
    }

    /**
     * Get system statistics
     * @returns {Promise<Object>} System statistics
     */
    async getStats() {
        return await this._request('GET', '/stats');
    }
}

/**
 * Custom error class for API errors
 */
class BroCodeAPIError extends Error {
    constructor(message, statusCode, details) {
        super(message);
        this.name = 'BroCodeAPIError';
        this.statusCode = statusCode;
        this.details = details;
    }

    toString() {
        return `BroCodeAPIError(${this.statusCode}): ${this.message}`;
    }
}

/**
 * Create a BroCode SDK client
 * @param {string} apiKey - Your API key
 * @param {string} baseUrl - Base URL of API
 * @returns {BroCodeSDK} SDK client instance
 */
function createClient(apiKey, baseUrl = 'http://localhost:5000') {
    return new BroCodeSDK(apiKey, baseUrl);
}

// Export for Node.js/CommonJS
if (typeof module !== 'undefined' && module.exports) {
    module.exports = {
        BroCodeSDK,
        BroCodeAPIError,
        createClient
    };
}

// Export for ES6 modules
if (typeof window !== 'undefined') {
    window.BroCodeSDK = BroCodeSDK;
    window.BroCodeAPIError = BroCodeAPIError;
    window.createBroCodeClient = createClient;
}

// Example usage
/*
// Initialize client
const client = new BroCodeSDK(
  'bk_your_api_key_here',
  'http://localhost:5000'
);

// Health check
try {
  const health = await client.healthCheck();
  console.log('API Status:', health.status);
} catch (error) {
  console.error('Error:', error.message);
}

// Register user
try {
  const keystrokes = [
    { key: 'a', press_time: 100, release_time: 150 },
    { key: 'b', press_time: 200, release_time: 250 }
  ];
  
  const result = await client.registerUser(
    'john_doe',
    'english',
    keystrokes,
    'john@example.com'
  );
  
  console.log('Registration:', result);
} catch (error) {
  console.error('Error:', error.message);
}

// Get trust score
try {
  const score = await client.getTrustScore('john_doe');
  console.log('Trust Score:', score.trust_score);
} catch (error) {
  console.error('Error:', error.message);
}
*/
